### PR TITLE
[Enhancement] add const partition column for hdfs scanner

### PR DESCRIPTION
**Pull request by letian-jiang**
## What I'm doing:

Reduce the cost of building partition columns, which are of the same value in hive.

before
image

after
image

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/column/binary_column.cpp [hunk 1]
```diff
@@ -62,6 +63,24 @@ void BinaryColumnBase<T>::append(const Column& src, size_t offset, size_t count)
 
 template <typename T>
 void BinaryColumnBase<T>::append_selective(const Column& src, const uint32_t* indexes, uint32_t from, uint32_t size) {
+    if (src.is_constant()) {
```
Comment (letian-jiang) [inline be/src/column/binary_column.cpp:66]: To Support `ConstColumn` would fail to cast in line 84.

Also, this is a quick path to handle `ConstColumn`.

Comment (satanson) [inline be/src/column/binary_column.cpp:66]: it seems nullable column, const null column, ordinary column processing branches are missing!
and constant column dictation is illegal, use statment as follows!!
if(!src.is_only_null() && src.is_constant()) {
...
}

File: be/src/column/binary_column.cpp [hunk 2]
```diff
@@ -62,6 +63,24 @@ void BinaryColumnBase<T>::append(const Column& src, size_t offset, size_t count)
...
+        auto& const_column = down_cast<const ConstColumn&>(src);
+        auto& data_column = const_column.data_column();
+        auto concrete_data_column = down_cast<const BinaryColumnBase<T>*>(data_column.get());
+        auto length = concrete_data_column->get_offset()[1] - concrete_data_column->get_offset()[0];
+        auto bytes = concrete_data_column->get_bytes();
+
```
Comment (satanson) [inline be/src/column/binary_column.cpp:72]: too tedious!

```cpp
        auto& const_column = down_cast<const ConstColumn&>(src);
        auto& data_column = const_column.data_column();
        auto concrete_data_column = down_cast<const BinaryColumnBase<T>*>(data_column.get());
        auto length = concrete_data_column->get_offset()[1] - concrete_data_column->get_offset()[0];
        auto bytes = concrete_data_column->get_bytes();
```
change to 
```cpp
auto value = ColumnHelper::get_constant_value<Slice>(src);
auto length = value.size;
...
```

File: be/src/exec/hdfs_scanner.cpp [hunk 3]
```diff
@@ -441,24 +441,9 @@ void HdfsScannerContext::append_or_update_partition_column_to_chunk(ChunkPtr* ch
...
-        }
-
-        if (ck->is_slot_exist(slot_desc->id())) {
-            ck->update_column(std::move(chunk_part_column), slot_desc->id());
-        } else {
-            ck->append_column(std::move(chunk_part_column), slot_desc->id());
-        }
+        auto partition_value = partition_values[i]->clone();
+        partition_value->resize(row_count);
+        ck->append_or_update_column(std::move(partition_value), slot_desc->id());
```
Comment (zombee0) [inline be/src/exec/hdfs_scanner.cpp:446]: const_null_column's type has conflict with chunk's schema.

File: be/src/column/binary_column.cpp (hunk 1 above)
Comment (letian-jiang) [reply be/src/column/binary_column.cpp]: fixed

File: be/src/column/binary_column.cpp (hunk 2 above)
Comment (letian-jiang) [reply be/src/column/binary_column.cpp:70]: done

**Review by satanson — DISMISSED** [review]

**Comment by letian-jiang** [comment]
another issue on exchange 
image

**Comment by letian-jiang** [comment]
Closed as the const column will be unfold in subsequent operators.

**Pull request closed without merging**